aws.codedeploy | application, deployment-group and deployment

PratMis opened the pull request:
Closes #6781 

This PR addresses 2 use cases wrt CodeDeploy:
1. Delete an application if it's linked to Github(since CodeDeploy doesn't support github enterprise yet)
2. Stop an active deployment if it's linked to Github or S3.

thisisshi commented on c7n/resources/code.py:
```diff
@@ -265,3 +265,70 @@ def process(self, resources):
+class DeleteApplication(BaseAction):
+
+    schema = type_schema('delete')
+    permissions = ('codepipeline:DeletePipeline',)
```
permissions seem wrong here

thisisshi commented on c7n/resources/code.py:
```diff
@@ -265,3 +265,70 @@ def process(self, resources):
+       Roll back is set to True by default
+    """
+
+    schema = type_schema('stop', autorollbackenabled={"type": "boolean"})
```
can you set the default here in the jsonschema as well

thisisshi commented on c7n/resources/code.py:
```diff
@@ -265,3 +265,70 @@ def process(self, resources):
+                self.manager.retry(client.stop_deployment,
+                      deploymentId=r['deploymentId'],
+                      autoRollbackEnabled=self.data.get('autorollbackenabled', True))
+            except client.exceptions.DeploymentAlreadyCompletedException:
```
probably also want to continue on `DeploymentDoesNotExistException`

thisisshi reviewed (commented):
To handle the use case outlined in the issue you posted you would also need a related resource filter to handle any deployments from a codedeploy repo from github or non-whitelisted s3 bucket as well. overall looks ok to me, some minor comments below

PratMis commented:
Thanks for the feedback @thisisshi. Currently we're trying to only restrict to using AppSec in conjunction with CodeDeploy but yes, you're right re: whitelisting of S3 buckets we would need a Related Resource filter. Happy to add that in the next PR. I have addressed your comments.

kapilt commented on c7n/resources/code.py:
```diff
@@ -265,3 +265,71 @@ def process(self, resources):
+        for r in resources:
+            try:
+                self.manager.retry(client.delete_application, applicationName=r['applicationName'])
+            except client.exceptions.InvalidApplicationNameException:
```
let's add ApplicationDoesNotExistException

PratMis replied:
Addressed

kapilt commented on c7n/resources/code.py:
```diff
@@ -265,3 +265,71 @@ def process(self, resources):
+        name = id = 'deploymentId'
+        # couldn't find a real cloudformation type
+        cfn_type = None
+        arn_type = ""
```
deploymentgroup appears to be the closest arn type https://docs.aws.amazon.com/codedeploy/latest/userguide/auth-and-access-control-permissions-reference.html

kapilt replied:
more specifically a deployment seems to always be qualified by both its application and deployment group. i'm curious what a deployment specifies here if it wasn't started with an explicit deployment group, if its a magic 'default' value for deployment group name/id. in some sense we're effectively skipping the optional parent, but perhaps thats okay (ie ancestry in the api appears to be app -> group -> deploy, but we're omitting group and going through all the data plane of deployments.

kapilt replied:
it feels like some of this would enforcement would be better targeted to deployment group instead of the data plane on deployment to prevent mid deployment issues.

PratMis replied:
Hello @kapilt , thanks for the feedback. I did some research on CodeDeploy and here's what I found:
1. To create a deployment group, it only requires the application name and deployment group name and my understanding is that `targetRevision` that we get back as part of [get_deployment_group](https://boto3.amazonaws.com/v1/documentation/api/latest/reference/services/codedeploy.html#CodeDeploy.Client.get_deployment_group) api call inherits it from the actual deployment(which can change based on the actual deployment). I did not have to specify those details while creating a Deployment Group.
2. The [get_deployment_group](https://boto3.amazonaws.com/v1/documentation/api/latest/reference/services/codedeploy.html#CodeDeploy.Client.get_deployment_group) api needs applicationName and DeploymentGroupName as part of required fields in the api which makes me think that it would have to be a child resource with Application as parent resulting in not having the luxury to have custodian monitor deployment groups in event based(CloudTrail) mode. Hence, I added a new deployment resource type that would monitor deployments in event based mode.
3.  I agree that we could add a DeploymentGroup Resource type later, but was wondering if we could get a cleaner api before we do that. Not sure if that's going to happen :)

Happy to hear your thoughts!

kapilt commented on c7n/resources/code.py:
```diff
@@ -265,3 +265,71 @@ def process(self, resources):
+        id = 'applicationId'
+        name = 'applicationName'
+        date = 'createTime'
+        arn_type = ""
```
arn_type = "application"

kapilt commented on c7n/resources/code.py:
```diff
@@ -265,3 +265,71 @@ def process(self, resources):
+
+    class resource_type(TypeInfo):
+        service = 'codedeploy'
+        enum_spec = ('list_deployments', 'deployments', None)
```
this should have a third param on static query to only include deployments not in a terminal state.

PratMis commented:
@kapilt I went ahead and added a `codedeploy-deploymentgroup` resource. Interestingly, deployment group inherits `targetRevision` from the most recent deployment. I was able to trigger two back to back deployments with different targetRevisions and found only the latest revision to be stored. I think we may still need a `codedeploy-deployment` resource to monitor deployments in cloudtrail mode. Please let me know your thoughts!! Thanks

PratMis commented:
I went ahead and removed `stop` action on deployments. Kept the resource itself since it would be beneficial to get to know about a deployment that didn't come from a whitelisted sources. Deployment Groups inherit `targetRevision` from deployments but only the latest revision. Having a CloudTrail mode policy on deployments would instantly notify about any non-compliance.

kapilt commented on c7n/resources/code.py:
```diff
@@ -265,3 +268,149 @@ def process(self, resources):
+
+    @property
+    def generate_arn(self):
+        if self._generate_arn is None:
```
why did this method need to be overriden?

kapilt replied:
i went ahead and updated the pr on this, i did note the need for name for id and different arn separate as driving some of this.

PratMis replied:
Thanks @kapilt, looks a lot cleaner

kapilt commented on c7n/resources/code.py:
```diff
@@ -265,3 +268,149 @@ def process(self, resources):
+        date = 'createTime'
+
+
+@query.sources.register('describe-deployment-group')
```
resource specific overrides of describe sources don't need registration anymore, they can be defined be defined as a map on the resource, to just override describe key.

kapilt replied:
which it looks like your already doing, so can just remove source registration here.

kapilt commented:
last consideration was just on resource naming, if we want to try and simplify application -> app  .. and deploymentgroup -> group.. thoughts?

PratMis commented:
> last consideration was just on resource naming, if we want to try and simplify application -> app .. and deploymentgroup -> group.. thoughts?

Yep, I can make that change

kapilt reviewed (approved):
Lgtm
